File: Coursera-Machine-Learning/Matlab-Code/machine-learning-ex1/ex1/gradientDescentMulti.m
function [theta, J_history] = gradientDescentMulti(X, y, theta, alpha, num_iters)
%GRADIENTDESCENTMULTI Performs gradient descent to learn theta
%   theta = GRADIENTDESCENTMULTI(x, y, theta, alpha, num_iters) updates theta by
%   taking num_iters gradient steps with learning rate alpha

% Initialize some useful values
m = length(y); % number of training examples
J_history = zeros(num_iters, 1);

for iter = 1:num_iters

    % ====================== YOUR CODE HERE ======================
    % Instructions: Perform a single gradient step on the parameter vector
    %               theta. 
    %
    % Hint: While debugging, it can be useful to print out the values
    %       of the cost function (computeCostMulti) and gradient here.
    %

% Compute hypothesis
h = X * theta;

% Compute error
error_vector = h - y;

% Compute gradient
grad_J = (1 / m) *  (X' * error_vector);

% Update Params
theta = theta - alpha * grad_J;









    % ============================================================

    % Save the cost J in every iteration    
    J_history(iter) = computeCostMulti(X, y, theta);

end

end

File: Coursera-Machine-Learning/Python-Code/ex1/gradientDescentMulti.py
import numpy as np
from computeCostMulti import computeCostMulti

def gradientDescentMulti(X, y, theta, alpha, num_iters):
    m = len(y) # number of training examples
    J_history = np.zeros((num_iters, 1)) # Using visualize gradient of cost function
    
    # Perform a single gradient step on the parameter vector theta
    for i in range(num_iters):
        # Compute gradient of cost function
        h = X.dot(theta)
        err_vec = h - y
        grad_J = (1 / m) * X.T.dot(err_vec)
        
        theta = theta - alpha * grad_J # update theta
        
        # Save the cost J in every iteration
        J_history[i] = computeCostMulti(X, y, theta)
        
    return [theta, J_history]